torch/testing/_internal/composite_compliance.py

```diff
@@ -0,0 +1,172 @@
+
+# TODO: move this into library proper
+@contextlib.contextmanager
+def no_dispatch() -> Iterator[None]:
+    guard = torch._C._DisableTorchDispatch()  # type: ignore[attr-defined]
+    try:
+        yield
+    finally:
+        del guard
+
+def check_metadata_consistency(wrapper_tensor):
+    if not isinstance(wrapper_tensor, CompositeCompliantTensor):
+        return
+    elem = wrapper_tensor.elem
+    if wrapper_tensor.shape != elem.shape:
+        raise RuntimeError(
+            "This operator is not CompositeImplicitAutograd compliant: the "
+            "shape of the tensor was modified directly without "
+            "going through the PyTorch dispatcher.")
+    if wrapper_tensor.dtype != elem.dtype:
+        raise RuntimeError(
+            "This operator is not CompositeImplicitAutograd compliant: the "
+            "dtype of the tensor was modified directly without "
+            "going through the PyTorch dispatcher.")
+
+def is_view_fn(func):
+    return func.__name__ in {
+        'as_strided',
+        'detach',
+        'diagonal',
+        'expand',
+        'expand_as',
+        'movedim',
+        'narrow',
+        'permute',
+        'select',
+        'squeeze',
+        'transpose',
+        't',
+        'real',
+        'imag',
+        'view_as_real',
+        'view_as_complex',
+        'unflatten',
+        'unfold',
+        'unsqueeze',
+        'view',
+        'view_as',
+        'unbind',
+        'split',
+        'split_with_sizes',
+        'vsplit',
+        'hsplit',
+        'tensor_split',
+        'chunk',
+        'swapaxes',
+        'slice',
+        '_reshape_alias',
+        '_unsafe_view',
+        '_conj',
+        'alias',
+    }
+
+def is_inplace_view_fn(func):
+    return func.__name__ in {
+        'squeeze_',
+        'unsqueeze_',
+        'transpose_',
+        't_',
+    }
+
+class CompositeCompliantTensor(torch.Tensor):
+    elem: torch.Tensor
+
+    __slots__ = ['elem']
+
+    @staticmethod
+    def __new__(cls, elem, *args, **kwargs):
+        # The storage of CompositeCompliantTensor should never be used directly
+        # by a CompositeImplicitAutograd operation; if the CompositeImplicitAutograd
+        # operator attempts to read from the storage without dispatching then it'll
+        # raise a RuntimeError due to it being a meta storage.
+        r = torch.Tensor._make_wrapper_subclass(  # type: ignore[attr-defined]
+            cls, elem.size(),
+            dtype=elem.dtype, layout=elem.layout,
+            device=elem.device, requires_grad=elem.requires_grad)
+        r.elem = elem
+        return r
+
+    def __repr__(self):
+        return f"CompositeCompliantTensor({self.elem})"
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+        def unwrap(e):
+            return e.elem if isinstance(e, CompositeCompliantTensor) else e
+
+        def wrap(e):
+            return CompositeCompliantTensor(e) if isinstance(e, torch.Tensor) else e
```

Comment:
A fridge thought: shouldn't pytree only ever call wrap on Tensors...

Reply:
pytree APIs (like tree_map) are useful for non-tensors. For example, `vmap(f, in_dims)(*args)`: the `in_dims` is allowed to be a pytree of the same "structure" as args and what we want to do there is "zip" the in_dims with the `args`.